Test non-standard date and time formats

labkey-danield opened the pull request:
#### Rationale
Test automation for non-standard date and time formats. The tests are in NonStandardDateAndTimeFormatTest.java.
Consolidated various test helpers that set date and time fields. Change various helpers, UI components, other tests, etc... to use the enums defined in the base setting page for the format not a string (not a string).

Added a page to set formats at the page level (FolderFormatsPage).

Added APIContainerHelper.setDateAndTimeFormats. This allows setting of non-standard formats to projects and folders.

#### Related Pull Requests
- https://github.com/LabKey/limsModules/pull/888

#### Changes
- Consolidate and clean up helpers.
- Add back tests that had been commented out.
- Added NonStandardDateAndTimeFormatTest.java
- Updated tests and other code to use helpers from LKS tests.
- Added check for format warnings to domain designer and settings pages.
- APIContainerHelper.setDateAndTimeFormats

labkey-danield committed:
Consolidate helpers that interact with Date, Time and DateTime fields.
Change helpers to take enum values and not strings.

labkey-danield committed:
Re-enable ProjectSettingsTest.testInjection.

labkey-danield committed:
Fix error in DomainFieldRow.setFormat

labkey-danield committed:
Change the none time format to have a format value of "<none>".
Uncomment a test that fails with older versions of FireFox.

labkey-danield committed:
Use variable and remove reference to string value.

labkey-danield committed:
Merge branch 'develop' into fb_testNonStandardDateAndTimeFormats

labkey-danield committed:
Merge branch 'develop' into fb_testNonStandardDateAndTimeFormats

labkey-danield committed:
Testing non-standard formats in list and projects.

labkey-danield committed:
Added test for Data Sets.
Added check for Site Validation report.

labkey-danield committed:
Merge branch 'develop' into fb_testNonStandardDateAndTimeFormats

labkey-danield committed:
Add comments and descriptions doe each test.

labkey-danield committed:
Merge branch 'develop' into fb_testNonStandardDateAndTimeFormats

labkey-danield committed:
Adding tests for child folder and site settings.

labkey-danield committed:
Merge branch 'develop' into fb_testNonStandardDateAndTimeFormats

labkey-danield committed:
Add test for site and subfolder settings.
Add a FolderFormats page.
Added a helper in the test that cna be used to validate the format settings on a site, project or folder page.

labkey-danield committed:
Cleaning up code and comments.

labkey-danield committed:
Merge branch 'release24.11-SNAPSHOT' into 24.11_fb_testNonStandardDateAndTimeFormats

labkey-danield committed:
Merge branch 'release24.11-SNAPSHOT' into 24.11_fb_testNonStandardDateAndTimeFormats

labkey-danield committed:
Merge branch 'release24.11-SNAPSHOT' into 24.11_fb_testNonStandardDateAndTimeFormats

labkey-danield committed:
Merge branch 'release24.11-SNAPSHOT' into 24.11_fb_testNonStandardDateAndTimeFormats

labkey-danield committed:
Validate using 'Date', 'Time' and 'DateTime' as format values.

labkey-danield committed:
Merge branch 'release24.11-SNAPSHOT' into 24.11_fb_testNonStandardDateAndTimeFormats

labkey-danield committed:
Merge branch 'release24.11-SNAPSHOT' into 24.11_fb_testNonStandardDateAndTimeFormats

labkey-danield committed:
Merge branch 'release24.11-SNAPSHOT' into 24.11_fb_testNonStandardDateAndTimeFormats

labkey-danield committed:
Merge branch 'release24.11-SNAPSHOT' into 24.11_fb_testNonStandardDateAndTimeFormats

labkey-danield committed:
Merge branch 'release24.11-SNAPSHOT' into 24.11_fb_testNonStandardDateAndTimeFormats

labkey-danield committed:
Some minor clean up.

XingY commented on src/org/labkey/test/components/domain/DomainFormPanel.java:
```diff
@@ -125,7 +127,36 @@ private DomainFormPanel editField(DomainFieldRow fieldRow, FieldDefinition field
+            {
+                String format = fieldDefinition.getFormat().trim();
+
+                int index = format.indexOf(" ");
```
I don't think this works with `MMMM dd yyyy` date time format?

labkey-danield replied:
You are right.
I have changed the code to this:
```
                String format = fieldDefinition.getFormat().trim();
                int index = format.lastIndexOf(" ");

                if (format.substring(index + 1).contains(":"))
                {
                    fieldRow.setDateTimeFormat(
                            DATE_FORMAT.get(format.substring(0, index)),
                            TIME_FORMAT.get(format.substring(index + 1)));

```
If the "last part" of the format string contains a ':' the it has a time value. If it doesn't then it is a date only format.

XingY commented on src/org/labkey/test/pages/core/admin/BaseSettingsPage.java:
```diff
@@ -351,12 +398,19 @@ public enum DATE_FORMAT
         ddMMMyy("ddMMMyy"),
+        MMddyyyy("MM/dd/yyyy"),
+        MM_dd_yyyy("MM-dd-yyyy"),
+        MMMM_dd_yyyy("MMMM dd yyyy"),
```
is `MMMM_dd_yyyy` used in any test? It's kinda special in that it contains space. I think it's good to check setting date/datetime field that uses this type and verify it's persisted / parsed correctly.

labkey-danield replied:
I have changed a couple of tests in LKS and LKSM to use `MMMM_dd_yyyy`

XingY commented on src/org/labkey/test/tests/NonStandardDateAndTimeFormatTest.java:
```diff
@@ -0,0 +1,1334 @@
+
+    // Private helper to check the Site Validation report. Checks to see if the report should, or should not, contain
+    // the list of warnings.
+    private void validateSiteValidationReport(String scope, List<String> expectedWarnings, boolean shouldContain)
```
Is there test that checks meta formats "date", "datetime" and "time" won't show up on the report?

labkey-danield replied:
The NonStandardDateAndTimeFormatTest.testTextFormatValueOfDate uses the `DATE, TIME, DATETIME and none` values. I've added a call to check that the site validation report does not flag them.

labkey-danield committed:
In DomainFormPanel.editField fix how a DateTime format is parsed.
Use the date format that has spaces in it in a few tests.
In NonStandardDateAndTimeFormatTest.testTextFormatValueOfDate check that using DATE, TIME and DATETIME do not get flagged by the site validation report.
In SMNonStandardDateTimes.testAdminPage add a check for setting the formats back to standard values.
